tools/codegen/gen.py

```diff
@@ -334,10 +337,39 @@ def go(f: NativeFunction) -> Optional[str]:
 
         assert target is Target.DEFINITION
 
-        def generate_defn(sig: CppSignature) -> str:
+        def generate_defn(sig: CppSignature, faithful: bool) -> str:
             dispatcher_sig = DispatcherSignature.from_schema(f.func)
 
-            dispatcher_exprs = dispatcher.cpparguments_exprs(sig.argument_packs())
+            # Partitions a sequence of argument packs into (out_args, args)
+            def partition_argument_packs(args: Sequence[CppArgumentPack]) -> Tuple[List[CppArgumentPack], List[CppArgumentPack]]:
+                return partition(args, lambda arg: isinstance(arg, CppSingleArgumentPack) and arg.is_out_argument)
+
+            # We're generating the dispatcher expressions from the C++ API argument list
+            # but while the argument list was generated from the JIT schema, the arguments
+            # can have been reordered. We need to undo that and bring arguments back into
+            # the JIT schema order to call into the dispatcher.
+            # There are 4 cases:
+            # 1. faithful signature for c10-full ops
+            #    This doesn't need us to do anything since C++ API argument order and dispatcher argument order match
+            # 2. non-faithful signature for non-c10-full ops
+            #    This also doesn't need us to do anything since they match as well
+            # 3. faithful signature for non-c10-full ops
+            #    faithful signature has out arguments in the back, the dispatcher for non-c10-full ops takes them
+            #    in the front. We need to reorder them to the front.
+            # 4. non-faithful signature for c10-full ops
+            #    non-faithful signature has out arguments in the front, the dispatcher for c10-full ops takes them
+            #    in the back. We need to reorder them to the back.
+            # Cases 1, 2 don't need special handling and cases 3, 4 are detected and handled in the following code block.
+            # Note that cases 2, 3 will die as soon as all ops are c10-full but 1, 4 will remain.
```

Comment:
This seems off - if we need to undo an upstream change, it means this logic should live upstream and produce different artifacts for this code to consume. Off the top of my head an obvious approach to do would be to produce a different `CppSignature` for the faithful and non-faithful cases. Is there an issue with doing that?